File: scripts/verify-plan.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { spawnSync } from "node:child_process";

type AtlasSource = {
  id: string;
  title: string;
  kind: string;
  priority: string;
  primary_use: string;
  access: string;
  urls: string[];
  expected_artifacts: string[];
  verifier: string;
};

type AtlasManifest = {
  schema_version: string;
  generated_from: string;
  sources: AtlasSource[];
};

const root = new URL("..", import.meta.url).pathname.replace(/\/$/, "");
const wikiRoot = "/Users/jasonlaster/src/projects/diana-tnbc/obsidian/wiki/omics";

const errors: string[] = [];
const warnings: string[] = [];

function requireFile(path: string) {
  if (!existsSync(path)) {
    errors.push(`Missing required file: ${path}`);
  }
}

function readProjectFile(relativePath: string) {
  const fullPath = join(root, relativePath);
  requireFile(fullPath);
  return existsSync(fullPath) ? readFileSync(fullPath, "utf8") : "";
}

function commandVersion(command: string, args: string[]) {
  const result = spawnSync(command, args, { encoding: "utf8" });
  if (result.status !== 0) {
    return null;
  }
  return `${result.stdout}${result.stderr}`.trim().split("\n")[0] ?? "";
}

const requiredProjectFiles = [
  "README.md",
  "docs/PROJECT_PLAN.md",
  "docs/SOURCE_MAP.md",
  "docs/WIKI_SOURCE_SUMMARY.md",
  "manifests/validation_atlases.json"
];

for (const file of requiredProjectFiles) {
  requireFile(join(root, file));
}

const requiredWikiFiles = [
  "index.md",
  "findings-overview.md",
  "derived-findings.md",
  "analysis-workflows.md",
  "validation-atlases.md",
  "partner-questions.md"
];

for (const file of requiredWikiFiles) {
  requireFile(join(wikiRoot, file));
}

const plan = readProjectFile("docs/PROJECT_PLAN.md");
for (let i = 0; i <= 7; i += 1) {
  if (!plan.includes(`Milestone ${i}`)) {
    errors.push(`PROJECT_PLAN.md is missing Milestone ${i}.`);
  }
}

for (const required of ["Verifier", "HRD", "Bun", "Python", "TCGA-BRCA", "CHORD", "scarHRD"]) {
  if (!plan.includes(required)) {
    errors.push(`PROJECT_PLAN.md is missing required term: ${required}`);
  }
}

const sourceMap = readProjectFile("docs/SOURCE_MAP.md");
for (const requiredUrl of [
  "https://gdc.cancer.gov/about-data/publications/brca_2012",
  "https://docs.cbioportal.org/downloads/",
  "https://xena.ucsc.edu/",
  "https://www.nature.com/articles/nm.4292",
  "https://github.com/UMCUGenetics/CHORD"
]) {
  if (!sourceMap.includes(requiredUrl)) {
    errors.push(`SOURCE_MAP.md is missing required URL: ${requiredUrl}`);
  }
}

let manifest: AtlasManifest | null = null;
try {
  manifest = JSON.parse(readProjectFile("manifests/validation_atlases.json")) as AtlasManifest;
} catch (error) {
  errors.push(`validation_atlases.json is not valid JSON: ${String(error)}`);
}

if (manifest) {
  if (!manifest.schema_version) {
    errors.push("Manifest is missing schema_version.");
  }
  if (!Array.isArray(manifest.sources) || manifest.sources.length < 10) {
    errors.push("Manifest should include at least 10 researched sources/tools.");
  }

  const ids = new Set<string>();
  let hasPhase1 = false;
  let hasTool = false;
  let hasDataset = false;

  for (const source of manifest.sources) {
    if (ids.has(source.id)) {
      errors.push(`Duplicate manifest source id: ${source.id}`);
    }
    ids.add(source.id);

    for (const field of ["id", "title", "kind", "priority", "primary_use", "access", "verifier"] as const) {
      if (!source[field]) {
        errors.push(`Manifest source ${source.id || "(missing id)"} is missing ${field}.`);
      }
    }

    if (!Array.isArray(source.urls) || source.urls.length === 0) {
      errors.push(`Manifest source ${source.id} must include at least one URL.`);
    }

    if (!Array.isArray(source.expected_artifacts) || source.expected_artifacts.length === 0) {
      errors.push(`Manifest source ${source.id} must include expected artifacts.`);
    }

    hasPhase1 ||= source.priority.includes("phase-1");
    hasTool ||= source.kind === "tool";
    hasDataset ||= source.kind === "dataset";
  }

  if (!hasPhase1) {
    errors.push("Manifest has no phase-1 source.");
  }
  if (!hasTool) {
    errors.push("Manifest has no tool sources.");
  }
  if (!hasDataset) {
    errors.push("Manifest has no dataset sources.");
  }
}

const pythonVersion = commandVersion("python3", ["--version"]);
if (!pythonVersion) {
  errors.push("python3 is not available.");
}

const rVersion = commandVersion("R", ["--version"]);
if (!rVersion) {
  warnings.push("R is not available locally; R-native tools should use a container or later R setup.");
}

if (!process.versions.bun) {
  errors.push("This verifier should be run with Bun.");
}

async function checkUrls() {
  if (!manifest || process.env.CHECK_URLS !== "1") {
    return;
  }

  for (const source of manifest.sources) {
    for (const url of source.urls) {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 10000);
      try {
        let response = await fetch(url, { method: "HEAD", signal: controller.signal });
        if (response.status === 405 || response.status === 403) {
          response = await fetch(url, { method: "GET", signal: controller.signal });
        }
        if (response.status >= 400) {
          warnings.push(`URL returned ${response.status}: ${url}`);
        }
      } catch (error) {
        warnings.push(`URL check failed for ${url}: ${String(error)}`);
      } finally {
        clearTimeout(timeout);
      }
    }
  }
}

await checkUrls();

for (const warning of warnings) {
  console.warn(`warning: ${warning}`);
}

if (errors.length > 0) {
  for (const error of errors) {
    console.error(`error: ${error}`);
  }
  process.exit(1);
}

console.log("Plan verification passed.");
if (pythonVersion) {
  console.log(`Python: ${pythonVersion}`);
}
console.log(`Bun: ${process.versions.bun}`);

File: py/src/diana_omics/commands/verify_plan.py
from __future__ import annotations

import json
import os
import subprocess
import sys
import urllib.error
import urllib.request
from pathlib import Path
from typing import Optional

from ..paths import ROOT, WIKI_ROOT


def command_version(command: str, args: list[str]) -> Optional[str]:
    try:
        result = subprocess.run([command] + args, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=False)
    except FileNotFoundError:
        return None
    if result.returncode != 0:
        return None
    output = f"{result.stdout}{result.stderr}".strip().splitlines()
    return output[0] if output else ""


def check_url(url: str) -> Optional[str]:
    for method in ("HEAD", "GET"):
        request = urllib.request.Request(url, method=method)
        try:
            with urllib.request.urlopen(request, timeout=10) as response:
                if response.status < 400:
                    return None
                if response.status not in (403, 405):
                    return f"URL returned {response.status}: {url}"
        except urllib.error.HTTPError as error:
            if error.code not in (403, 405):
                return f"URL returned {error.code}: {url}"
        except Exception as error:
            return f"URL check failed for {url}: {error}"
    return None


def main() -> None:
    errors: list[str] = []
    warnings: list[str] = []

    def require_file(path: Path) -> None:
        if not path.exists():
            errors.append(f"Missing required file: {path}")

    def read_project_file(relative_path: str) -> str:
        full_path = ROOT / relative_path
        require_file(full_path)
        return full_path.read_text(encoding="utf-8") if full_path.exists() else ""

    for file in [
        "README.md",
        "docs/PROJECT_PLAN.md",
        "docs/SOURCE_MAP.md",
        "docs/WIKI_SOURCE_SUMMARY.md",
        "manifests/validation_atlases.json",
    ]:
        require_file(ROOT / file)
    for file in [
        "index.md",
        "findings-overview.md",
        "derived-findings.md",
        "analysis-workflows.md",
        "validation-atlases.md",
        "partner-questions.md",
    ]:
        require_file(WIKI_ROOT / file)

    plan = read_project_file("docs/PROJECT_PLAN.md")
    for index in range(8):
        if f"Milestone {index}" not in plan:
            errors.append(f"PROJECT_PLAN.md is missing Milestone {index}.")
    for required in ["Verifier", "HRD", "Bun", "Python", "TCGA-BRCA", "CHORD", "scarHRD"]:
        if required not in plan:
            errors.append(f"PROJECT_PLAN.md is missing required term: {required}")

    source_map = read_project_file("docs/SOURCE_MAP.md")
    for required_url in [
        "https://gdc.cancer.gov/about-data/publications/brca_2012",
        "https://docs.cbioportal.org/downloads/",
        "https://xena.ucsc.edu/",
        "https://www.nature.com/articles/nm.4292",
        "https://github.com/UMCUGenetics/CHORD",
    ]:
        if required_url not in source_map:
            errors.append(f"SOURCE_MAP.md is missing required URL: {required_url}")

    manifest = None
    try:
        manifest = json.loads(read_project_file("manifests/validation_atlases.json"))
    except Exception as error:
        errors.append(f"validation_atlases.json is not valid JSON: {error}")

    if manifest:
        if not manifest.get("schema_version"):
            errors.append("Manifest is missing schema_version.")
        sources = manifest.get("sources")
        if not isinstance(sources, list) or len(sources) < 10:
            errors.append("Manifest should include at least 10 researched sources/tools.")
            sources = []
        ids = set()
        has_phase1 = has_tool = has_dataset = False
        for source in sources:
            if source.get("id") in ids:
                errors.append(f"Duplicate manifest source id: {source.get('id')}")
            ids.add(source.get("id"))
            for field in ["id", "title", "kind", "priority", "primary_use", "access", "verifier"]:
                if not source.get(field):
                    errors.append(f"Manifest source {source.get('id') or '(missing id)'} is missing {field}.")
            if not source.get("urls"):
                errors.append(f"Manifest source {source.get('id')} must include at least one URL.")
            if not source.get("expected_artifacts"):
                errors.append(f"Manifest source {source.get('id')} must include expected artifacts.")
            has_phase1 = has_phase1 or "phase-1" in source.get("priority", "")
            has_tool = has_tool or source.get("kind") == "tool"
            has_dataset = has_dataset or source.get("kind") == "dataset"
        if not has_phase1:
            errors.append("Manifest has no phase-1 source.")
        if not has_tool:
            errors.append("Manifest has no tool sources.")
        if not has_dataset:
            errors.append("Manifest has no dataset sources.")

    python_version = command_version("python3", ["--version"])
    if not python_version:
        errors.append("python3 is not available.")
    if not command_version("R", ["--version"]):
        warnings.append("R is not available locally; R-native tools should use a container or later R setup.")
    bun_version = command_version("bun", ["--version"])
    if not bun_version:
        warnings.append("Bun is not available on PATH; package.json task aliases require Bun, but Python commands can still run directly.")

    if manifest and os.environ.get("CHECK_URLS") == "1":
        for source in manifest.get("sources", []):
            for url in source.get("urls", []):
                warning = check_url(url)
                if warning:
                    warnings.append(warning)

    for warning in warnings:
        print(f"warning: {warning}", file=sys.stderr)
    if errors:
        for message in errors:
            print(f"error: {message}", file=sys.stderr)
        raise SystemExit(1)
    print("Plan verification passed.")
    if python_version:
        print(f"Python: {python_version}")
    if bun_version:
        print(f"Bun: {bun_version}")


if __name__ == "__main__":
    main()
